File: seal/signal.py
"""SEAL signal: rToken vs last cash close, with event-aware side selection."""

from typing import Optional


def compute_spread(rtoken_price: float, cash_close: float) -> float:
    if cash_close <= 0:
        raise ValueError("cash_close must be positive")
    return rtoken_price / cash_close - 1.0


def is_event_window(date: str, event_dates: frozenset[str]) -> bool:
    return date in event_dates


def parent_side(spread: float, threshold: float, is_event: bool) -> Optional[str]:
    """Return "long", "short", or None if the spread doesn't clear the threshold.

    Event window: trade with the move (spread direction = side direction).
    No event: fade the spike (empty-book overreaction assumed to mean-revert).
    """
    if abs(spread) < threshold:
        return None
    moved_up = spread > 0
    trade_with_move = moved_up if is_event else not moved_up
    return "long" if trade_with_move else "short"

File: backend/src/signal.ts
/** Direct TS port of seal/signal.py — same names, same behavior. Pinned to
 * the Python implementation by test/golden.test.ts. */

export type Side = "long" | "short" | null;

export function computeSpread(rtokenPrice: number, cashClose: number): number {
  if (cashClose <= 0) {
    throw new Error("cashClose must be positive");
  }
  return rtokenPrice / cashClose - 1.0;
}

export function isEventWindow(date: string, eventDates: ReadonlySet<string>): boolean {
  return eventDates.has(date);
}

export function parentSide(spread: number, threshold: number, isEvent: boolean): Side {
  if (Math.abs(spread) < threshold) {
    return null;
  }
  const movedUp = spread > 0;
  const tradeWithMove = isEvent ? movedUp : !movedUp;
  return tradeWithMove ? "long" : "short";
}
